### DISCUSSION TITLE
ModuleNotFoundError: No module named 'imagehash'?

### DISCUSSION THREAD

**Discussion by Rapid1898-code** [pyinstaller/pyinstaller#7090] [Help]
Hello - i have written a python-code which is using the imagehash module

```
from PIL import Image
import imagehash
import requests
img = "https://i.imgur.com/5bGzZi7.jpg"
resp = requests.get(img, stream=True)
ihashImg = imagehash.average_hash(Image.open(resp.raw)) 
print(ihashImg)
```

This works fine when i run the python-code in vscode.
But when i compile with pyinstaller (version 5.4.1) and try to create the windiows-exe with eg.

```
pyinstaller --onefile --hidden-import=imagehash test1.py
pyinstaller --onefile test1.py
```

With both variants i get this error when running the executeable:

```
Traceback (most recent call last):
  File "test1.py", line 2, in <module>
    import imagehash
ModuleNotFoundError: No module named 'imagehash'
[24964] Failed to execute script test1
```

Any ideas why this is not working?
Is it necessary to use some other parameters?

**Comment by rokm** [comment]
> Any ideas why this is not working?
> Is it necessary to use some other parameters?

If a direct import from your entry-point script is not found at runtime, chances are very high that it cannot be found during the freezing process, typically because you have PyInstaller installed in a different python environment (virtual or base) than you usually use to run your code.

**Comment by Rapid1898-code** [comment]
Hello - pyinstaller is also installed in the venv which i am using for program start.

**Comment by rokm** [comment]
What if you run pyinstaller via `python -m PyInstaller` instead of using the `pyinstaller` shim? The latter might (also) be installed in some other environment (which you can check with `where pyinstaller`)...

**Comment by Rapid1898-code** [comment]
When i write this
`python -m pyinstaller --onefile test2.py`
he tells me for whatever reason
`C:\Users\Polzi\Documents\DEV\.venv\test\Scripts\python.exe: No module named pyinstaller`

But when i do pip list (within the activated venv) i get this:
```
pyinstaller               5.4.1
pyinstaller-hooks-contrib 2022.10
```

How is this possible?

When i do where pyinstaller i get:
```
C:\Users\Polzi\Documents\DEV\.venv\test\Scripts\pyinstaller.exe
C:\Users\Polzi\AppData\Local\Programs\Python\Python39\Scripts\pyinstaller.exe
```

**Comment by bwoodsend** [comment]
PyInstaller the module name is case sensitive. You need to use `python -m PyInstaller`.

**Comment by Rapid1898-code** [comment]
Thanks a lot - i am not sure, but somehow i get this to work.
(i deleted the pyinstaller in the base installment and then it seems to work).
What is generally the best practice for this?
Install the pyinstaller in every venv - but not in the base?

**Reply by rokm** [reply]
> What is generally the best practice for this? Install the pyinstaller in every venv - but not in the base?

Yes. Or use `python -m PyInstaller` to avoid confusion around multiple `pyinstaller.exe` shims being in `PATH`.

**Comment by Rapid1898-code** [comment]
ok - thanks a lot for you help - you can close this question pls
